lib/core/src/Cardano/Wallet.hs

```diff
@@ -2087,6 +2104,54 @@ guardHardIndex ix =
     then throwE $ ErrIndexOutOfBound minBound maxBound ix
     else pure (Index $ getDerivationIndex ix)
 
+updateCosigner
+    :: forall ctx s k n.
+        ( HasDBLayer s k ctx
+        , s ~ SharedState n k
+        , MkKeyFingerprint k (Proxy n, k 'AddressK CC.XPub)
+        , MkKeyFingerprint k Address
+        , SoftDerivation k
+        , Typeable n
+        , WalletKey k
+        )
+    => ctx
+    -> WalletId
+    -> k 'AccountK XPub
+    -> Cosigner
+    -> CredentialType
+    -> ExceptT ErrAddCosignerKey IO ()
+updateCosigner ctx wid accXPub cosigner cred = db & \DBLayer{..} -> do
+    mapExceptT atomically $ do
+        cp <- withExceptT ErrAddCosignerKeyNoSuchWallet $ withNoSuchWallet wid $
+              readCheckpoint (PrimaryKey wid)
+        let s@(SharedState _ st) = getState cp
+        case st of
+            PendingFields (SharedStatePending _ pT dTM _) -> do
+                when (cred == Delegation && isNothing dTM)$
+                    throwE ErrAddCosignerKeyNoDelegationTemplate
+
+                when (cred == Payment && isCosignerMissing pT) $
+                    throwE $ ErrAddCosignerKeyNoSuchCosigner cosigner Payment
+                when (cred == Delegation && isCosignerMissing (fromJust dTM)) $
+                    throwE $ ErrAddCosignerKeyNoSuchCosigner cosigner Delegation
+
+                when (cred == Payment && isKeyAlreadyPresent pT) $
+                    throwE $ ErrAddCosignerKeyAlreadyPresentKey Payment
+                when (cred == Delegation && isKeyAlreadyPresent (fromJust dTM)) $
+                    throwE $ ErrAddCosignerKeyAlreadyPresentKey Delegation
+
+                let s' = addCosignerAccXPub accXPub cosigner cred s
```

Comment:
It seems to me that pretty much all guards above should actually be part of the `addCosignerAccXPub` function since they pretty much lay down the pre-conditions for this function. 

Having the validation logic separated from this function may bite us later. If we expect that any call to this function does indeed check these pre-conditions, then they would be better captured within the function that should fail (e.g `Either`) when misused.

Reply:
done